alphagov/paas-cf
# [#111428758] Generate certs for cloudfoundry in each deployment

[Auto-generate SSLs for all internal components](https://www.pivotaltracker.com/story/show/111428758)

## What

We want to generate the certificates for all the components, signed by a CA which is trusted by all the related VMs. We will generate the certificates for all the current SSL certificates used in the deployment, as the ones hardcoded [here](https://github.com/alphagov/paas-cf/blob/7b649d96e6b730bfaf9980717c3ed08bbc20899d/manifests/cf-manifest/outputs/cf-ssl-certificates.yml) 
 
## Dependencies

This PR must be merged first https://github.com/alphagov/paas-docker-cloudfoundry-tools/pull/16
 and the a dockerhub automated build container added and used in this PR.

## Context

This PR should be reviewed in this context:

 * I want to generate a CA certificate to sign the rest of certificates and that is setup as a [trusted certificate](https://bosh.io/docs/trusted-certs.html) in all deployed VMs.
 * I want to generate and use the certificates for the gorouter with the deployment entry domain.
 * I want to to pass the generated CA to consul and BBS services.
 * I want to generate the signing key for UAA tokens.
 * I want to remove any hardcoded certificate or key.

Additionally:
 * I want to make it easy to add new YAML stubs to the bosh and CF manifest.
 * I want to make it easy to generate and add new certificates, even new CAs.
 * I want all tests to pass.

## Notes

 * In the future we should use a real SSL certificate for the router, but for the time being is OK to use one generated by us.
 * In this PR we are, *again* abusing of the script s3init.  Because that, the generation jobs are not connected to the consumer. To solve this, we suggest  that in the future we refactor the certificates as resources, with this logic for `in`: *"If the certificate is in a S3 bucket and will not expire, download it from S3. If not, generate it and upload it to the S3 bucket"*

## How to test this?

 1. Deploy the system as usual. Note that **you must redeploy bosh** as, a new config is required, and  redeploy all CF vms **from scratch** so they get the trusted certificate.
     * If the deployment is OK, the certificate work for BBS, consul.
 2. Run `smoke_tests`, they should pass.
 3. Run `acceptance_tests`, but note that some tests as flacky and might fail.

You can also check if the gorouter certificate is accepted by any other VM by SSHing  to one server and doing `https://api.${DEPLOY_ENV}.cf.paas.alphagov.co.uk/v2/info`

Example, from one cell:

```
bosh_z2lictl6z@52cfb1f0-bd9b-4b90-9463-2a8ac748da70:~$ curl https://api.michael.cf.paas.alphagov.co.uk/v2/info
{"name":"vcap","build":"2222","support":"http://support.cloudfoundry.com","version":2,"description":"Cloud Foundry sponsored by Pivotal","authorization_endpoint":"https://login.michael.cf.paas.alphagov.co.uk","token_endpoint":"https://uaa.michael.cf.paas.alphagov.co.uk","min_cli_version":null,"min_recommended_cli_version":null,"api_version":"2.43.0","app_ssh_endpoint":"ssh.michael.cf.paas.alphagov.co.uk:2222","app_ssh_host_key_fingerprint":null,"app_ssh_oauth_client":"ssh-proxy","routing_endpoint":"https://api.michael.cf.paas.alphagov.co.uk/routing","logging_endpoint":"wss://loggregator.michael.cf.paas.alphagov.co.uk:443","doppler_logging_endpoint":"wss://doppler.michael.cf.paas.alphagov.co.uk:443"}
```

But from our personal computer:
```
$ curl https://api.michael.cf.paas.alphagov.co.uk/v2/info
curl: (60) SSL certificate problem: unable to get local issuer certificate
```

# who?

Anyone but @keymon or @mtekel
